sql,coord: beginnings of ALTER CLUSTER

benesch opened the pull request:
ALTER CLUSTER is a possibly short-lived command to add and remove
replicas of clusters. Specifically, the syntax of CREATE CLUSTER ...
REMOTE is changed to

    CREATE CLUSTER <name> [REMOTE <name> (<host>, <host> ...) ...]

where multiple REMOTE clauses may now be specified to name multiple
replicas. A nearly identical ALTER CLUSTER command

    ALTER CLUSTER <name> [REMOTE <name> (<host>, <host> ...) ...]

can we used to wholesale replace the remote replica set of a cluster.
Replicas are identified by the user provided name. After alteration, any
new replicas are added and any unnamed replicas are removed. Changing
the host specification of an existing replica is not permitted.

Forthcoming work will make the concept of a replica more explicitly part
of the system (e.g., by allowing virtual and managed replicas). But for
now this unblocks the active replication work.

### Motivation

  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered. (Unstable feature; tests later.)

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

frankmcsherry commented on src/coord/src/coord.rs:
```diff
@@ -2021,6 +2025,18 @@ impl Coordinator {
+    ) -> Result<ExecuteResponse, CoordError> {
+        // TODO(benesch): update catalog with new configuration.
+
+        // TODO(mcsherry?): frob the controller with the diff of the compute
+        // instance configuration (`plan.config`).
```
Do we need to do anything here? We could just ship the new instance plan at the controller and have it sort things out (i.e. do the diff there, as it is the source of truth of what is installed).

frankmcsherry commented on src/dataflow-types/src/client.rs:
```diff
@@ -71,8 +71,7 @@ pub enum InstanceConfig {
-        hosts: Vec<String>,
+        replicas: Vec<RemoteReplica>,
         /// Logging configuration.
         logging: Option<LoggingConfig>,
```
One thing this raises is that the logging stuff isn't going to be well replicated. Peeks and stuff might be ok, but TAIL against any of the introspection sources will be undefined.

frankmcsherry reviewed (commented):
Looks like a good start. I bet we could just feed the intended config at the compute instance controller and ask it to do the differencing, if we trust that any names that stay put mean the replica is unchanged too (probably not a legit assumption, but we could get started with that).

benesch commented:
Sweet! I'm going to push up a fix for the failed tests.

benesch commented:
Great! I fixed the test failures locally and marked for automerge. No substantive changes, so I squashed into a single commit with a proper commit message.

`test/cluster` has a working example of using `ALTER CLUSTER`. Well, the test is still flaky for reasons unrelated to `ALTER CLUSTER`, but if you manage to avoid that flake, the addition of the second replica seems to work!
